albatross/eigen/serializable_ldlt.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_EIGEN_SERIALIZABLE_LDLT_H
+#define ALBATROSS_EIGEN_SERIALIZABLE_LDLT_H
+
+#include "Eigen/Cholesky"
+#include "Eigen/Dense"
+#include "cereal/cereal.hpp"
+#include <math.h>
+
+namespace Eigen {
+
+template <class Archive, typename _Scalar, int _Rows, int _Cols>
+inline void save_lower_triangle(Archive &archive,
+                                const Eigen::Matrix<_Scalar, _Rows, _Cols> &v) {
+  cereal::size_type rows = static_cast<cereal::size_type>(v.rows());
+  cereal::size_type storage_size = (rows * rows + rows) / 2;
+  archive(cereal::make_size_tag(storage_size));
+  for (cereal::size_type i = 0; i < rows; i++) {
+    for (cereal::size_type j = 0; j <= i; j++) {
+      archive(v(i, j));
+    }
+  }
+}
+
+template <class Archive, typename _Scalar, int _Rows, int _Cols>
+inline void load_lower_triangle(Archive &archive,
+                                Eigen::Matrix<_Scalar, _Rows, _Cols> &v) {
+  cereal::size_type storage_size;
+  archive(cereal::make_size_tag(storage_size));
+  // TODO: understand why the storage size upon load is always augmented by two.
+  storage_size -= 2;
+  // We assume the matrix is square and compute the number of rows from the
+  // storage size.
+  double drows = (std::sqrt(static_cast<double>(storage_size * 8 + 1)) - 1) / 2;
+  cereal::size_type rows = static_cast<cereal::size_type>(drows);
+  v.resize(rows, rows);
+  for (cereal::size_type i = 0; i < rows; i++) {
+    for (cereal::size_type j = 0; j <= i; j++) {
+      archive(v(i, j));
+    }
+  }
+}
+
+class SerializableLDLT : public LDLT<MatrixXd, Lower> {
+public:
+  SerializableLDLT() : LDLT<MatrixXd, Lower>(){};
+
+  SerializableLDLT(const LDLT<MatrixXd, Lower> &ldlt)
+      // Can we get around copying here?
+      : LDLT<MatrixXd, Lower>(ldlt){};
```

Comment:
If this stores a unique pointer we could `move` it in which would save the copy?

Reply:
If I understand correct, the problem with that is that it would require changing the way `Eigen::LDLT` works.  As it stands all the storage is done in the base class.  We might be able to do something like instantiate an empty `Eigen::LDLT` then point the internal members to pointers that are passed in ... but that feels wrong.

Reply:
Yeah probably not worth it.